Move handling of secure locking into logind

AdrianVovk opened the pull request:
Depends on #31841 

I'd like to get this into v256, to make sure that we can use this in GNOME 47 (releasing September 2024)

Here's a quick summary / todo-list of the changes:
- [x] Introduced a new logind `SecureLock()`API to loosen the coupling between homed and other components (systemd-sleep, the desktop environment, etc)
- [x] Introduced functionality that warns clients about an imminent `SecureLock()`, and lets them delay it for a little while. It's pretty much the same thing as logind's delay inhibitor + `PrepareForSleep()` mechanism. This lets session services prepare for a secure lock by wiping secrets out of memory (think: your ssh-agent), closing sensitive FDs, and unsubscribing from the bus (since they'll be frozen)
- [x] Made homed a backend for the `SecureLock()` API
- [x] Moved homed's "please_suspend" logic into logind, again to loosen the coupling between desktop environment and homed
- [x] Deprecated the various features in homed















<!-- devel-freezer = {"comment-id":"2083568757","freezing-tag":"v256-rc1"} -->

AdrianVovk committed:
logind: Add SecureLock()

This is a special lock operation that wipes sensitive data out of
memory, makes files inaccessible, and freezes the user session (i.e.
homed's lock feature).

This is implemented in a backend-agnostic manner, to loosen the coupling
between systemd components and to allow desktop environments to support
this feature no matter what service is managing the user.

This impl. also supports warning clients about an imminent secure lock,
and letting them delay the lock for a short while so that they can
prepare themselves. This will allow session services to additionally
wipe their own secrets from memory to further improve the security of a
secure lock.

AdrianVovk committed:
logind: Add SecureLockUsers() method

Securely locks all the users that logind is tracking

AdrianVovk committed:
homed: Deprecate LockAllHomes()

Since it has moved into logind. We mark LockAllHomes() as deprecated and
calling it just proxies through into logind's equivalent API

AdrianVovk committed:
homed: Implement SecureLock() backend for logind

This implements the io.systemd.SecureLockBackend Varlink service in
homed for logind's SecureLock() functionality

AdrianVovk committed:
loginctl: Add secure-lock-user verb

To expose the secure-locking behavior via the CLI, so that it can be
conveniently activated.

AdrianVovk committed:
pam_systemd: Allow sessions to disable secure lock

Some sessions are going to be incompatible w/ secure locking (i.e. if
the desktop environment doesn't have a reauthentication mechanism, or if
logged in on a TTY). Let's completely disable the functionality if any
session belonging to the user reports that it doesn't support secure
locking

AdrianVovk committed:
homed: Remove vestigial please_suspend logic

This was moved into logind, and does nothing anyways since
LockAllHomes() was stubbed out

AdrianVovk committed:
Update docs about homed lock-on-suspend

We will move all of this into logind. Plus, we will have a lot more
flexibility about how/where locking happens, so let's update the docs to
reflect all of the new functionality.

AdrianVovk committed:
logind: Take over automatic secure locking

Before, we had logind handling delay inhibitors, then starting sd-sleep,
which would call back into homed to handle secure locking.

Now that we've moved secure locking into logind, logind would start
sd-sleep and sd-sleep would call back into logind. Instead of
implementing it like this, we just make logind handle secure locking
itself before it starts sd-sleep. This is better layering and makes it
easier to reason about what's happening immediately before a suspend.

poettering commented on src/home/homectl.c:
```diff
@@ -2059,6 +2059,9 @@ static int lock_home(int argc, char *argv[], void *userdata) {
         int r, ret = 0;
 
+        log_warning("Locking directly via homed is a low-level operation and is thus not recommended.");
+        log_warning("Please use 'loginctl secure-lock-user' instead.");
```
merge into one string with a \n in the middle.

(so that this becomes one journal event, not two)

poettering commented on man/org.freedesktop.login1.xml:
```diff
@@ -992,6 +992,14 @@ node /org/freedesktop/login1/user/_1000 {
     methods:
       Terminate();
       Kill(in  i signal_number);
+      SecureLock();
```
adding a concept such as a "SecureLock" might be OK, but it kinda irks me that our screen lock concept is currently so weird, i.e. we only fire a signal, that's it. ideally we'd track the screen lock status, and allow clients to wait for the screen lock to be activated.

it seems to me that for the secure lock we'd provide features we should also provide for the screen lock.

generally i think distinguishing between two levels of lock, i.e. screen lock and secure lock makes sense. I just wish they would expose more similar semantics.

AdrianVovk replied:
I agree that normal screen locks should be improved. Unfourtunately, I'm not sure if there's a great way to do that in a backwards-compatible way. And anyway, I'm just not sure that is necessarily actionable for this PR. Screen locks are per-session concepts, but secure locks are per-user concepts

What do you want me to implement re: screen locks in this PR?

Or if you're speaking in general, I could add a TODO entry here

poettering replied:
it's fine to not hack anything up for screen locks right now, all i am asking for to keep the concept in mind.

my original request was kinda in the hope we could have a common set of functions that just takes an enum or flags, asking for one of the two kinds.

but i figure this doesn't actually make much sense, since the screen locks lock session objects, while the "secure locks" lock users as a whole.

hence maybe ignore me.

poettering commented on man/org.freedesktop.login1.xml:
```diff
@@ -992,6 +992,14 @@ node /org/freedesktop/login1/user/_1000 {
       Kill(in  i signal_number);
+      SecureLock();
+      DelaySecureLock(out h delay_fd);
+      InhibitAutoSecureLock(out h inhibitor_fd);
```
hmm, can't we make this part of the usual inhibitor logic?

poettering replied:
ah, i guess you want this per-user while the existing inhibitor logic is per-system.

so i'd really like to keep the door open to expose a screen lock in similar ways one day. hence I think this should take a bit mask, and one bit should be for the secure lock thing.

(the per-system inhibitor locks take this weird ":" separated string. it should really be a bitmask though. The ":" thing was a historical mistake we should not repeat).

AdrianVovk replied:
Right, this is per-user but normal inhibitors are system-wide.

I can't reuse this for normal screen locks either, because screen locks are per-session, which is again not per-user.

If you want a more flexible inhibitors API anyway, I have a few questions:
- do you want the "who"/"why" fields?
- will these need to be listed in any way?
- there are some invalid configurations, like delaying auto-secure-lock. I would also like to prevent callers from blocking secure lock in general (letting clients opt the user out of a security feature even if the user manually triggers it seems like a bad idea...). What to do about them?

poettering replied:
i think the who/why fields make sense. at worst they are a bit of noise, but at best they are nice debugging.

listing them eventually would make sense, but no need to cover that initially.

returning errors when something shall not be allowed seems fine.

poettering replied:
and better to prohibit too much initially (and losen things later maybe) than the reverse.

AdrianVovk replied:
I decided to just let callers fully inhibit secure locking. Maybe there's a use-case. I do prevent callers from delaying things that don't make sense to delay, however.

poettering commented on man/org.freedesktop.login1.xml:
```diff
@@ -1074,13 +1098,63 @@ node /org/freedesktop/login1/user/_1000 {
+      <function>TriggerSecureLock()</function> signal. In response, the back-end should do whatever it needs
+      to securely lock the user, and then close the FD returned by this method to signal that it is done. The
+      back-end can also close the FD early to signal that secure locking has become unavailable for this user.
+      </para>
```
i'd prefer if this would be done the other way round. i.e. homed should maybe just announce it's capability in some user record field it adds to the "state" section of the record or so.

and then maybe extend the userdbd varlink API with a lock function. if logind wants to lock homed it would asynchronously fire a varlink call to that.

poettering replied:
i mean, this is ultimately sort of a feature of the userdb record provider, and hence should probably be requested the way you usually talk to a userdb provider.

this would then mean homed never has to call into logind for this. Instead logind would do a varlink call into homed when it needs to via the generic userdb api.

AdrianVovk replied:
Overall I'm not against this, and your concerns make sense to me. However, so far the generic userdb API has been entirely read-only to the user state. Layering-wise it makes total sense, but API wise it doesn't IMO. Maybe putting it there is the wrong place.

How about we introduce a separate varlink API for the secure-lock-backend (with a signal method to lock a user), and put it somewhere that logind can consume it. For example: `/run/systemd/secure-lock-backend/$SERVICE` where `$SERVICE` corresponds to the service in the user record. So let's say logind wants to secure-lock a user managed by homed: it'll look up the service in the user record, find `io.systemd.Home`, then connect to `/run/systemd/secure-lock-backend/io.systemd.Home` and call `LockUser(username)`.

On the user-record side, we'd document that `status.state` can be `locked` so that clients like logind can rely on it.

Then logind will have everything it needs:
- Check whether or not secure lock is possible: check whether or not `/run/systemd/secure-lock-backend/$SERVICE` exists
- Check whether or not the user is already secure locked: check `status.state` in the user record for `locked`
- Activate the secure lock for the user: Call the private varlink API

Thoughts?

poettering replied:
people have asked for update apis in the userdb api since day one. I figure eventually we have to give them something, hence maybe sooner or later userdb should just have a generic api "UpdateUserRecord" or so. in other words: the fact that userdb stuff is read-only right now is not necessarily a design decision but just effect because noone added writability yet...

hence I am not sure it's really worth it trying to keep the secure lock stuff separate. 

that said, I am not opposed to doing this on a separate interface if you want, but please drop the "backend" string from the socket path. /run/systemd/secure-lock/<service> or /run/systemd/userdb-secure-lock/ sounds ok?

poettering replied:
i think it might make sense to add a field canSecureLock or so to the "status" section of the user record, that homed adds in according to its chosen backend (after all the simple dir-based stuff doesn't know the concept).

AdrianVovk replied:
> people have asked for update apis in the userdb api since day one

Eh I think this is a bad idea. Different backends will have wildly different semantics about how they'll be updated. Different clients will need to implement these different semantics anyways. Might as well communicate directly with the backend

In other words, I'm somewhat against this. IMO the userdb stuff should be read-only, and if you want to write you need to look at `service` and figure out if you know how to write for that `service` backend.

> i think it might make sense to add a field canSecureLock

100% makes sense :+1:

poettering reviewed (changes requested):
So, see my comments. I think tracking/triggering the the lock state via logind makes sense, such a secure lock *is* indeed somewhat related to a screen lock. But i'd prefer that if we think about the secure lock we also keep the screen lock in mind and that we can model them both in similar fashion eventually (even if this is not implemented like this right now).

And i am very concerned about the layering, i.e. what is client to what.

And I'd prefer if logind would talk to homed via varlink, i.e. how it generally resolves user entries already. in an entirely generic way.

AdrianVovk committed:
homed: Operation: Support varlink calls

Previously, Operation supported replying to pending DBus calls. Now, it
also supports replying to pending Varlink calls. This allows us to
initiate operations via Varlink, which will be used later to implement
the secure-lock backend

AdrianVovk committed:
inhibitors: Add plumbing for per-user inhibitors

AdrianVovk committed:
user-record: Add canSecureLock field

To report whether or not the service providing a user record supports
secure-locking it

Also documents that `locked` is a known state of a user

AdrianVovk committed:
logind: Session: Don't claim we emit State changes

We don't emit them

AdrianVovk committed:
logind: Secure lock users that become inactive

If a user is inactive, that means that the user's sessions are
inaccessible from any seat. This will happen, for instance, if the
person at the seat presses the "Switch Users" button in their desktop
environment. In this case, we should secure-lock the now inaccessible
user.
